libpod/in_memory_state.go

```diff
@@ -0,0 +1,279 @@
+		if err != nil {
+			return errors.Wrapf(err, "error checking if container %s is present in pod %s", ctr.ID(), ctr.pod.ID())
+		} else if !hasCtr {
+			return errors.Wrapf(ErrNoSuchCtr, "container %s is not present in pod %s", ctr.ID(), ctr.pod.ID())
+		}
+	}
+
+	if err := s.ctrNameIndex.Reserve(ctr.Name(), ctr.ID()); err != nil {
+		return errors.Wrapf(err, "error registering container name %s", ctr.Name())
+	}
+
+	if err := s.ctrIDIndex.Add(ctr.ID()); err != nil {
+		s.ctrNameIndex.Release(ctr.Name())
+		return errors.Wrapf(err, "error registering container ID %s", ctr.ID())
+	}
+
+	s.containers[ctr.ID()] = ctr
+
+	return nil
+}
+
+// RemoveContainer removes a container from the state
+// The container will only be removed from the state, not from the pod the container belongs to
+func (s *InMemoryState) RemoveContainer(ctr *Container) error {
+	// Almost no validity checks are performed, to ensure we can kick
+	// misbehaving containers out of the state
+
+	_, ok := s.containers[ctr.ID()]
+	if !ok {
+		return errors.Wrapf(ErrNoSuchCtr, "no container exists in state with ID %s", ctr.ID())
+	}
+
+	if err := s.ctrIDIndex.Delete(ctr.ID()); err != nil {
+		return errors.Wrapf(err, "error removing container ID from index")
+	}
+	delete(s.containers, ctr.ID())
+	s.ctrNameIndex.Release(ctr.Name())
+
+	return nil
+}
+
+// GetAllContainers retrieves all containers from the state
+func (s *InMemoryState) GetAllContainers() ([]*Container, error) {
+	ctrs := make([]*Container, 0, len(s.containers))
+	for _, ctr := range s.containers {
+		ctrs = append(ctrs, ctr)
+	}
+
+	return ctrs, nil
+}
+
+// GetPod retrieves a pod from the state from its full ID
+func (s *InMemoryState) GetPod(id string) (*Pod, error) {
+	if id == "" {
+		return nil, ErrEmptyID
+	}
+
+	pod, ok := s.pods[id]
+	if !ok {
+		return nil, errors.Wrapf(ErrNoSuchPod, "no pod with id %s found", id)
+	}
+
+	return pod, nil
+}
+
+// LookupPod retrieves a pod from the state from a full or unique partial ID or
+// a full name
+func (s *InMemoryState) LookupPod(idOrName string) (*Pod, error) {
+	if idOrName == "" {
+		return nil, ErrEmptyID
+	}
+
+	fullID, err := s.podNameIndex.Get(idOrName)
+	if err != nil {
+		if err == registrar.ErrNameNotReserved {
+			// What was passed is not a name, assume it's an ID
+			fullID, err = s.podIDIndex.Get(idOrName)
+			if err != nil {
+				if err == truncindex.ErrNotExist {
+					return nil, errors.Wrapf(ErrNoSuchPod, "no pod found with name or ID %s", idOrName)
+				} else {
+					return nil, errors.Wrapf(err, "error performing truncindex lookup for ID %s", idOrName)
+				}
+			}
+		} else {
+			return nil, errors.Wrapf(err, "error performing registry lookup for ID %s", idOrName)
+		}
+	}
+
+	pod, ok := s.pods[fullID]
+	if !ok {
+		// This should never happen
+		return nil, errors.Wrapf(ErrInternal, "mismatch in pod ID registry and pod map for ID %s", fullID)
+	}
+
+	return pod, nil
+}
+
+// HasPod checks if a pod with the given ID is present in the state
```

Comment:
Same as above, I think there is no need for three calls, which really look to do the same thing.

Reply:
If we drop Has you'd have to parse the errors to determine if you actually got a ErrNoSuchPod or if it was an internal error in the state when you check for the presence of a pod/container

Reply:
I still think it is strange for a library to have three calls that can determine whether a container or pod exists.